### ISSUE TITLE
Fail to build engine for ARMM64/Aarch64 on Mac

### ISSUE THREAD

**Issue by galvesribeiro** [flutter/flutter#40513]
Hello folks!

I've followed the steps described on the [wiki](https://github.com/flutter/flutter/wiki/Compiling-the-engine) to get a regular build of flutter on Mac OSX and it builds just fine. 

From the same Mac machine, I would like to cross compile it to my target device (NVidia Jetson Nano - arm64) using the sysroot extracted from the the board itself and with clang/libcxxabi/libcxx I've built following @chinmaygarde [blog post for raspberry PI](https://medium.com/flutter/flutter-on-raspberry-pi-mostly-from-scratch-2824c5e7dcb1) and changed it to use our target device (i.e. aarch64-linux-gnu).

The `gn` command ended up as follow:

```
./flutter/tools/gn    \
     --clang                    \
    --target-sysroot /Volumes/Data/flutter-work/sdk/sysroot       \
    --target-toolchain /Volumes/Data/flutter-work/sdk/toolchain   \
    --target-triple aarch64-linux-gnu   \
    --linux-cpu arm64                     \
    --runtime-mode debug                \
    --no-lto                            \
    --target-os linux                   \
    --arm-float-abi hard                \
    --no-goma                           \
    --embedder-for-target   
```

And I'm ending up with this:

```
Generating GN files in: out/linux_debug_arm64
Generating Xcode projects took 79ms
Done. Made 353 targets from 176 files in 510ms
```

And this error:
```
$ ninja -C out/linux_debug_arm64 libflutter_engine.so -j48
ninja: Entering directory `out/linux_debug_arm64'
[1782/3448] CXX clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io.console_posix.o
FAILED: clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io.console_posix.o 
../../buildtools/linux-x64/clang/bin/clang++ -MD -MF clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io.console_posix.o.d -DDART_IO_SECURE_SOCKET_DISABLED -DUSE_OPENSSL=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DTARGET_ARCH_ARM64 -DNDEBUG -DDART_PRECOMPILER -DTARGET_OS_LINUX -I../../third_party/dart/runtime -I../../third_party -I../.. -Iclang_x64/gen -I../../third_party/libcxx/include -I../../third_party/libcxxabi/include -I../../third_party/dart/runtime/include -I../../third_party/zlib -fno-strict-aliasing -fstack-protector --param=ssp-buffer-size=4 -m64 -march=x86-64 -fPIC -pipe -pthread -fcolor-diagnostics -Wall -Wextra -Wendif-labels -Werror -Wno-missing-field-initializers -Wno-unused-parameter -fvisibility=hidden -Wheader-hygiene -Wno-implicit-int-float-conversion -Wstring-conversion -Wnewline-eof -Wthread-safety -O2 -fno-ident -fdata-sections -ffunction-sections -g0 -Werror -Wall -Wextra -Wno-unused-parameter -Wno-unused-private-field -Wnon-virtual-dtor -Wvla -Wno-conversion-null -Woverloaded-virtual -Wno-comments -g3 -ggdb3 -fno-rtti -fno-exceptions -Wimplicit-fallthrough -O3 -fvisibility-inlines-hidden -std=c++17 -fno-rtti -nostdinc++ -fno-exceptions   -c ../../third_party/dart/runtime/bin/console_posix.cc -o clang_x64/obj/third_party/dart/runtime/bin/gen_snapshot_dart_io.console_posix.o
/bin/sh: ../../buildtools/linux-x64/clang/bin/clang++: No such file or directory
[1805/3448] CXX obj/third_party/skia/src/sksl/skia.SkSLIRGenerator.o
ninja: build stopped: subcommand failed.
```

I know building to other devices is not a scenario really supported yet but I would appreciate any help on this issue.

So, basically to make it work, we need to build both the `host` and the `target` device parts but besides that error, I have no idea how to build the host part as there is no instructions what so ever on how to do so.

Can someone from google post a simple Wiki page (besides help on this issue) that show just the minimal _things_ required to build the flutter engine and the host part so people can just adapt it to their cases?

Thanks!

I really appreciate any help.

**Comment by chinmaygarde** [comment]
Unfortunately, due to the way the GN rules treat host vs target builds, this is currently only possible on a Linux host. If you don't have a Linux workstation handy, you should be able to use Docker for Mac.

About the `host` vs `target` device parts for your specific use case: Using the flags you just specified on a Linux host (or a Docker with Docker on Mac), you will get `libflutter_engine.so`, `icudtl.dat` and `embedder_engine.h`. These artifacts along with an embedder that uses the same will be used on the target board. Beside that, you will need to snapshot your Flutter application. This snapshot can be prepared by just calling `flutter build bundle` using the Flutter tools on the host (make sure the engine version the tools are using is the same as the artifacts you just built for your board).

> Can someone from google post a simple Wiki page (besides help on this issue) that show just the minimal things required to build the flutter engine and the host part so people can just adapt it to their cases?

I'll author something that describes the minimal things necessary. In the meantime, hopefully the guidance above is useful. Thanks.

**Comment by galvesribeiro** [comment]
Hello @chinmaygarde and thanks for the quick reply.

Hummm now I understand... The engine build steps are correct. Except they need to be made on linux box for the device part.

The host part can be just a default build as long as it uses the same version so it can br used by the tooling like vscode. Interesting...

I thought there would be some need to built a target specific tool for the host so it knows how to generate the snapshot for the target cpu...

Looking forward to see your example. Please let us know here :) thank you!

**Comment by galvesribeiro** [comment]
@chinmaygarde updates! Did the build on a Ubuntu 18.04.3 VM.

I had to add `"-Wno-implicit-float-conversion"` to the `cflags` on `clang/BUILD.gn` otherwise the build would fail on the `icu` build step on a long -> double cast. After add that the build works just fine.

Now when I look at the output, the lib and the icu are there, but the header is not:

![image](https://user-images.githubusercontent.com/4714040/64923244-6cbb3400-d7ae-11e9-952b-400d4a9844cb.png)

Do I need to do anything else?

**Comment by chinmaygarde** [comment]
Awesome. The header is just [embedder.h](https://github.com/flutter/engine/blob/5075172fe7f7858c2e1449e53092d34f23a3338e/shell/platform/embedder/embedder.h). I thought it was copied to the out directory by default but maybe not. You can pick up that header. That contains the API vended by libflutter_engine.so.

You have everything you need now. You can use the .so and header to write the embedder for your board. See a sample implementation for [GLFW here](https://gist.github.com/chinmaygarde/8abf44921f7d87f6da7bf026267c4792). Code documentation for the [embedder API is here](https://engine.chinmaygarde.com/embedder_8h.html). Thats the entire public API.

Excited to see progress and keep me posted if possible. Thanks.

**Comment by chinmaygarde** [comment]
About `-Wno-implicit-float-conversion`, can you paste the Clang version your toolchain is using? I don't know why our default toolchains don't need that. We should just fix ICU too TBH.

**Comment by galvesribeiro** [comment]
Hello @chinmaygarde here is the clang version:

```
./clang --version
clang version 10.0.0 
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/guto/flutter-work/sdk/toolchain/bin/.
 ```

I've just got whatever is on the tip of master and built it along libcxxabi and libcxx.

Thanks!

**Comment by galvesribeiro** [comment]
```
./clang --version
clang version 10.0.0 
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/guto/flutter-work/sdk/toolchain/bin/.
```
Thas is what the `clang --version` return to me on this build that caused the float cast error.

**Comment by galvesribeiro** [comment]
Hey @chinmaygarde got more progress.

For whatever reason if I build the device part on linux in a shared directory mounted in a linux container and then reconfigure the source to build the host part on OSX, I get a lot of OSError from the python tools. I guess that when we run `gclient sync` there are some git hooks that download OS-specific tools used by the `gn` build that don't work cross platform. I've synced on another directory and now I have both host and the target builds on my machine.

Now it comes to tooling... To avoid have the loads of objects and temp stuff from the build shared across all the team, I'd like to pack only the necessary for both the target and the host users.

So now I have:

- **Device**: libflutter_engine.so, icudtl.dat, embedder.h + embedder binary + app snapshot
- **Host**: ????

So the point is, what should I get from `out/host_debug` so I can pack and send to all the team so they can develop apps, build snapshots and send to the device?

This is what I have now:
![image](https://user-images.githubusercontent.com/4714040/65790964-1a2b3180-e137-11e9-86bb-797585176b07.png)

After we have both Host and Device parts solved, here are more questions:

1. Do you have idea how to tell VSCode to use the host build we just have?
2. How to attach the debugger to the "remote" linux device running the embedder we've just built?

Again, sorry for all those questions as they may be questions for anyone trying to embed flutter to other "non-conventional" targets but I had to ask as I can't find clear docs on those subjects yet.

Thank yoU!

**Comment by chinmaygarde** [comment]
You're right, `gclient sync` will download tools that are specific to the host platform.

If you haven't modified the engine sources at all, find the Git SHA of the Flutter Engine repo (the //flutter directory in your source root). You can then tell your team to just modify the [engine.version](https://github.com/flutter/flutter/blob/master/bin/internal/engine.version) file of the Flutter checkout to use this SHA. `flutter build bundle` and all other commands should use the correctly versioned artifacts from the cloud.

If OTOH, you have made modifications to the engine repo, you'll need to package `gen_snapshot`, `vm_outline_strong.dill`, `vm_platform_strong.dill`, `flutter_patched_sdk`, `frontend_serer.dart.snapshot.dart`. You can then tell the team to use the flutter tools with the `--local-engine` pointing to the the directory that contains these artifacts. I'll add the same to the wiki as well.

**Comment by galvesribeiro** [comment]
@chinmaygarde Thanks for the info.

So, since I didn't made any changes on the engine, all I need todo is to ask the team to change `${FLUTTER_DIR}/bin/internal/engine.version` on their release flutter to use the SHA of the engine checkout we built to the device and we are all set? If yes, that means I don't even need to build the host tools, right?

What about the other questions regarding the debugger/observatory and VSCode?

Again, thanks for the help! I really appreciate that.

**Comment by chinmaygarde** [comment]
> If yes, that means I don't even need to build the host tools, right?

Correct. You will be using the tools that Flutter puts in the Google Cloud buckets.

> What about the other questions regarding the debugger/observatory and VSCode?

When you launch the engine in JIT (the default `debug` mode), it should start the observatory automatically. This should print something like "Observatory running at http://localhost:4567/" in the console. If that port is forwarded to the development machine (SSH port forwarding should work fine), then you can use `flutter attach  --observatory-port 4567` to connect to the remotely running application. This should give development functionality like hot-reload, etc.. I am not sure about VSCode because I don't use that IDE myself but it should allow you to connect using the UI as well.

By default, the engine will pick a random observatory port, if you want to make it pick a consistent observatory port (so that port forwarding is easier), you can pass the `--observatory-port 9090` argument to the engine via [`FlutterProjectArgs::command_line_argv`](https://github.com/flutter/engine/blob/57f35d2763b046488ac8dd0409c21ac85865e320/shell/platform/embedder/embedder.h#L807) field.

I am writing a guide with all these details as I mentioned above. So please keep these question coming :) . Very helpful in understanding the decisions you have to make and the issues you run into while attempting an embedder implementation.

**Comment by galvesribeiro** [comment]
> Correct. You will be using the tools that Flutter puts in the Google Cloud buckets.

Awesome! That makes things wayyyy easier :) 

> When you launch the engine in JIT (the default debug mode), it should start the observatory automatically.

Yeah, I've built the target enngine with `--runtime-mode debug ` so it should run on JIT mode. 

> If that port is forwarded to the development machine (SSH port forwarding should work fine), then you can use flutter attach --observatory-port 4567 to connect to the remotely running application.

Yeah, SSH tunnel should work on that case.

> you can pass the --observatory-port 9090 argument to the engine via FlutterProjectArgs::command_line_argv field

Yeah, that makes things easier to use the tunnel indeed and would make most of the tools to _just work_ (TM). 

So, that setup would allow me to:

1. Run on JIT mode, using the snapshot of the app
2. Debug the target device dart code

However, I'm not sure if using the JIT/debug mode is very efficient. `flutter` cli is able to publish/pack for iOS/Android the required files AOT'ed... 

How could I make the same for our target device? I believe the flutter tool at some point will requires the target toolchain so it can generate native code. Is that possible? If yes, how?

> I am writing a guide with all these details as I mentioned above. So please keep these question coming :) . Very helpful in understanding the decisions you have to make and the issues you run into while attempting an embedder implementation.

Absolutely! :) I wish flutter had a slack channel or something where people can hang around to discuss things like that. Thanks!

**Comment by chinmaygarde** [comment]
> However, I'm not sure if using the JIT/debug mode is very efficient. 

Depends on the use case. You can use it for development and wire up AOT for deployment just like the mobile shells.

> flutter cli is able to publish/pack for iOS/Android the required files AOT'ed... How could I make the same for our target device? I believe the flutter tool at some point will requires the target toolchain so it can generate native code. Is that possible? If yes, how?

You can wire up AOT mode Dart execution in the custom Flutter Embedder by following the guide in the wiki
https://github.com/flutter/flutter/wiki/Custom-Flutter-Engine-Embedding-in-AOT-Mode.

> I believe the flutter tool at some point will requires the target toolchain so it can generate native code.

No. A native toolchain is not required to generate native AOT code. `gen_snapshot` has the chops to generate this. You already get that from either the cloud bucket or by locally building the host tools.

> I wish flutter had a slack channel or something where people can hang around to discuss things like that.

We hang out in a public Discord server https://github.com/flutter/flutter/wiki/Chat#chatting-on-discord. The #help channel should be a good place to start.

**Comment by galvesribeiro** [comment]
Cool! :)

Btw, just tried set the `engine.version` to the `git rev-parse HEAD` and I got this:

```
flutter doctor
Downloading android-arm-profile/darwin-x64 tools...                     
Download failed.
URL: https://storage.googleapis.com/flutter_infra/flutter/25b164b5f49fe4592a8cd8d6cd08440522b36a75/android-arm-profile/darwin-x64.zip
Error: 404 Not Found
```

I assume that is because I've built from master and not from a particular branch tag released to the buckets, correct? If yes, where can I find the matching tags/releases?

Thank you!

**Comment by chinmaygarde** [comment]
That is not a valid Flutter Engine commit https://github.com/flutter/engine/commit/25b164b5f49fe4592a8cd8d6cd08440522b36a75.

Are you sure you are taking a look at the SHA at the the //flutter directory of your **engine** checkout?

**Comment by chinmaygarde** [comment]
Also, if you are on ToT, it take a little bit of time for the bots to build all the artifacts for the specific commit. It is usually fine if you wait for a few mins. But you can follow the status of each engine commit here https://ci.chromium.org/p/flutter/g/engine/console

**Comment by galvesribeiro** [comment]
AHHHHH... I was getting from `engine` directory (the root one where `gclient sync` ran) and not from `engine/src/flutter` 😄 my bad. Got `b520d7fd4b6d66851219f3a7460884a27ee4e038` and that worked!

```
$ flutter doctor
Downloading android-arm-profile/darwin-x64 tools...                 1.0s
Downloading android-arm-release/darwin-x64 tools...                 0.8s
Downloading android-arm64-profile/darwin-x64 tools...               0.7s
Downloading android-arm64-release/darwin-x64 tools...               0.8s
Downloading android-x86 tools...                                    1.6s
Downloading android-x64 tools...                                    1.6s
Downloading android-arm tools...                                    1.2s
Downloading android-arm-profile tools...                            1.0s
Downloading android-arm-release tools...                            1.0s
Downloading android-arm64 tools...                                  1.2s
Downloading android-arm64-profile tools...                          1.6s
Downloading android-arm64-release tools...                          0.9s
Downloading package sky_engine...                                   0.6s
Downloading common tools...                                         1.1s
Downloading common tools...                                         1.6s
Downloading darwin-x64 tools...                                     2.8s
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.9.1+hotfix.2, on Mac OS X 10.14.6 18G95, locale en-BR)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 11.0)
[✓] Android Studio (version 3.4)
[✓] VS Code (version 1.38.1)
[✓] Connected device (1 available)

• No issues found!
```

**Comment by galvesribeiro** [comment]
> By default, the engine will pick a random observatory port, if you want to make it pick a consistent observatory port (so that port forwarding is easier), you can pass the --observatory-port 9090 argument to the engine via FlutterProjectArgs::command_line_argv field.

![image](https://user-images.githubusercontent.com/4714040/65821251-85453880-e209-11e9-8be3-0e7d726dbf39.png)

I guess it is not respecting the parameter passed:

> flutter: Observatory listening on http://127.0.0.1:45747/VeGn0PtOqQ0=/

Or am I missing something?

**Comment by chinmaygarde** [comment]
The first argument of the command line is interpreted as the executable name. So the engine is thinking that `--observatory-port` is the executable name and is confused. This is documented in the header https://github.com/flutter/engine/blob/d5b392bbb67c5ecebe2b45f087f305bd53727e42/shell/platform/embedder/embedder.h#L796

**Comment by hhk7734** [comment]
Hi, @chinmaygarde, @galvesribeiro 

Help me.
I had no experience using clang, so I ran various tests for a week.
Then I tried for a few days but it doesn't work and there are a lot of blockages.

I've tried it following [medium post](https://medium.com/flutter/flutter-on-raspberry-pi-mostly-from-scratch-2824c5e7dcb1)

```shell
mkdir -p $HOME/flutter-engine/sdk/toolchain \
&& mkdir -p $HOME/flutter-engine/sdk/sysroot \
&& mkdir -p $HOME/flutter-engine/engine

cd /media/hhk7734/rootfs \
&& sudo cp -r lib $HOME/flutter-engine/sdk/sysroot \
&& sudo cp -r opt $HOME/flutter-engine/sdk/sysroot \
&& sudo cp -r usr $HOME/flutter-engine/sdk/sysroot

sudo apt-get update \
&& sudo apt install gcc clang build-essential make \
    subversion texinfo bison flex gcc-multilib g++-multilib cmake \
&& sudo apt autoremove -y
```

I tried rsync command, but it wasn't finished due to recursive links. So, I use cp.

```shell
git clone --branch llvmorg-9.0.1 --depth=1 https://github.com/llvm/llvm-project.git $HOME/flutter-engine/llvm-project \
&& cd $HOME/flutter-engine/llvm-project

mkdir build_llvm && cd build_llvm \
&& cmake ../llvm \
    -DLLVM_ENABLE_PROJECTS=clang \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_INSTALL_PREFIX=$HOME/flutter-engine/sdk/toolchain \
    -DLLVM_DEFAULT_TARGET_TRIPLE=aarch64-linux-gnu \
    -DLLVM_TARGETS_TO_BUILD=AArch64

make -j$(expr $(expr $(nproc) \* 6) \/ 5) \
&& sudo make install
```

Added `-DLLVM_ENABLE_PROJECTS=clang`. Without it, I didn't find to install clang command.

```shell
git clone --depth=1 git://sourceware.org/git/binutils-gdb.git $HOME/flutter-engine/binutils-gdb \
&& cd $HOME/flutter-engine/binutils-gdb

./configure --prefix="$HOME/flutter-engine/sdk/toolchain" \
    --enable-gold \
    --enable-ld \
    --target=aarch64-linux-gnu

make -j$(expr $(expr $(nproc) \* 6) \/ 5) \
&& sudo make install
```

After the above, I tried the below commands. But failed.

```shell
cd $HOME/flutter-engine/llvm-project

mkdir build_libcxxabi && cd build_libcxxabi \
&& cmake ../libcxxabi \
    -DCMAKE_CROSSCOMPILING=True \
    -DCMAKE_SYSROOT=$HOME/flutter-engine/sdk/sysroot \
    -DLIBCXX_ENABLE_SHARED=False \
    -DCMAKE_INSTALL_PREFIX=$HOME/flutter-engine/sdk/toolchain \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_SYSTEM_NAME=Linux \
    -DCMAKE_SYSTEM_PROCESSOR=AArch64 \
    -DCMAKE_C_COMPILER=$HOME/flutter-engine/sdk/toolchain/bin/clang \
    -DCMAKE_CXX_COMPILER=$HOME/flutter-engine/sdk/toolchain/bin/clang++ \
    -DLLVM_TARGETS_TO_BUILD=AArch64 \
    -DLIBCXXABI_ENABLE_EXCEPTIONS=False
```

**error msg is**

```shell
-- Check for working CXX compiler: /home/hhk7734/flutter-engine/sdk/toolchain/bin/clang++ -- broken
CMake Error at /usr/share/cmake-3.10/Modules/CMakeTestCXXCompiler.cmake:45 (message):
  The C++ compiler

    "/home/hhk7734/flutter-engine/sdk/toolchain/bin/clang++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/hhk7734/flutter-engine/llvm-project/build_libcxxabi/CMakeFiles/CMakeTmp
    
    Run Build Command:"/home/hhk7734/flutter-engine/sdk/sysroot/usr/bin/make" "cmTC_485c2/fast"
    /home/hhk7734/flutter-engine/sdk/sysroot/usr/bin/make: 1: /home/hhk7734/flutter-engine/sdk/sysroot/usr/bin/make: Syntax error: "(" unexpected
```

I didn't find a solution. So, without it, I tried a flutter compilation.

```shell
 ./flutter/tools/gn \            
    --clang \
    --target-sysroot $HOME/flutter-engine/sdk/sysroot \
    --target-toolchain $HOME/flutter-engine/sdk/toolchain \
    --target-triple aarch64-linux-gnu \
    --linux-cpu arm64 \
    --runtime-mode debug \
    --embedder-for-target \
    --no-lto \
    --target-os linux \
    --arm-float-abi hard
```

But, failed with `/home/hhk7734/flutter-engine/sdk/toolchain/bin/aarch64-linux-gnu-ld: cannot find -lgcc_s`

Then, I added flag `--gcc-toolchain=/opt/toolchains/gcc-linaro-6.3.1-2017.02-x86_64_aarch64-linux-gnu/` to `linux_debug_arm64/toolchain.ninja` and `linux_debug_arm64/clang_x64/toolchain.ninja`.

Failed again. Besides this, I tried other things but failed.

Please help me. or recommend helpful resources.
I have a passion to study. :)

**Comment by hhk7734** [comment]
Some problems solved.
As I used cp command, some sysroot links were broken. I fixed it.

```shell
./flutter/tools/gn \
    --target-sysroot $HOME/flutter-engine/sdk/sysroot \
    --target-toolchain $HOME/flutter-engine/sdk/toolchain \
    --target-triple aarch64-linux-gnu \
    --linux-cpu arm64 \
    --runtime-mode debug \
    --embedder-for-target \
    --no-lto \
    --target-os linux \
    --arm-float-abi hard
```

```shell
ninja -C out/linux_debug_arm64
```

**Error**

```shell
ninja: Entering directory `out/linux_debug_arm64'
[3719/3722] SOLINK ./libflutter_engine.so
...
/home/hhk7734/flutter-engine/sdk/toolchain/bin/aarch64-linux-gnu-ld: obj/flutter/fml/platform/posix/fml.native_library_posix.o: in function `fml::NativeLibrary::NativeLibrary(char const*)':
native_library_posix.cc:(.text._ZN3fml13NativeLibraryC2EPKc+0x2c): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/home/hhk7734/flutter-engine/sdk/toolchain/bin/aarch64-linux-gnu-ld: /home/hhk7734/flutter-engine/sdk/sysroot/usr/lib/gcc/aarch64-linux-gnu/7.4.0/../../../aarch64-linux-gnu/libm.a(e_atan2.o): relocation R_AARCH64_ADR_PREL_PG_HI21 against symbol `__stack_chk_guard@@GLIBC_2.17' which may bind externally can not be used when making a shared object; recompile with -fPIC
/home/hhk7734/flutter-engine/sdk/toolchain/bin/aarch64-linux-gnu-ld: /home/hhk7734/flutter-engine/sdk/sysroot/usr/lib/gcc/aarch64-linux-gnu/7.4.0/../../../aarch64-linux-gnu/libm.a(e_atan2.o)(.text+0x18): unresolvable R_AARCH64_ADR_PREL_PG_HI21 relocation against symbol `__stack_chk_guard@@GLIBC_2.17'
/home/hhk7734/flutter-engine/sdk/toolchain/bin/aarch64-linux-gnu-ld: final link failed: bad value
clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

**Comment by hhk7734** [comment]
:) I got `libflutter_engine.so`, `icudtl.dat`, `flutter_embedder.h`.

**Comment by darshankawar** [comment]
@galvesribeiro 
Is this still a valid issue using latest Flutter version or has this been resolved ?

**Comment by pedromassangocode** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.
